### ISSUE TITLE
Harsh Pewter Peacock - `CrosschainDistributor._setToken` might become unusable due to dependency upon successful interaction with old token

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-06-tokensoft-judging#146]
Harsh Pewter Peacock

medium

# `CrosschainDistributor._setToken` might become unusable due to dependency upon successful interaction with old token

## Summary
The successful execution of `CrosschainDistributor._setToken` is dependent upon a successful decrease allowance call on the previous token. This will cause issues if call to previous token reverts or returns false. 

## Vulnerability Detail
The external `setToken` function internally invokes the `_setTotal` function.

```solidity
  function _setToken(IERC20 _token) internal virtual override nonReentrant onlyOwner {
    // decrease allowance on old token
    _allowConnext(0);

    // effect
    super._setToken(_token);

    // increase allowance on new token
    _allowConnext(total - claimed);
  }
```

The presence of `setToken` is for providing the ability to change `token` to airdrop owners in case when the old token starts behaving incorrectly (possibly due to misconfiguration, bug, exploit, etc). So that the old token can be replaced with a new token.

However this replacement mechanism itself depends upon a successful execution of oldToken.approve call.

In case the old token:
- reverts the `approve` call
- returns `false` on `approve` call
- is selfdestructed

The `setToken` call will also get reverted preventing owners from changing the `token`.

## Impact
Tokensoft airdrop protocol is meant to be used by other protocols for launching their token airdrops and the ability to change `token` in case of emergencies is an intended offering.

This report explains how this offering can be blocked which results in owners not able to change token. 

## Code Snippet
https://github.com/sherlock-audit/2023-06-tokensoft/blob/main/contracts/contracts/claim/abstract/CrosschainDistributor.sol#L47-L59

## Tool used

Manual Review

## Recommendation

Consider wrapping the first approve call to the old token in a `try {}` block so that unexpected response can be handled properly.

**Comment by akshaysrivastav** [comment]
Escalate for 10 USDC

Text here

**Comment by sherlock-admin2** [comment]
> Escalate for 10 USDC
> 
> Text here

The escalation could not be created because you are not exceeding the escalation threshold.

You can view the required number of additional valid issues/judging contest payouts in your Profile page,
in the [Sherlock webapp](https://app.sherlock.xyz/audits/).

**Comment by akshaysrivastav** [comment]
Escalate

Text here

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> Text here

The escalation could not be created because you are not exceeding the escalation threshold.

You can view the required number of additional valid issues/judging contest payouts in your Profile page,
in the [Sherlock webapp](https://app.sherlock.xyz/audits/).

**Comment by akshaysrivastav** [comment]
Further context on this report.

- The `setToken` function is intended to be used for replacing the existing `token`. 
- The only possible reason for which an admin would want to replace the existing `token` would be the case when the existing token starts behaving incorrectly (due to bug or hack). 
- The successful replacement of existing token depends upon the successful execution of `existingToken.approve` call.

So it is quite likely that since the existing token is buggy, any call to it (like the `approve` call) can fail which will prevent the replacement of token with a new token. Hence the Distributor contract will stay stuck with existing token forever with no way of replacement.

Also, this category of issues were accepted as `in-scope` by the sponsors, as per the [contest readme](https://github.com/sherlock-audit/2023-06-tokensoft/tree/main#q-in-case-of-external-protocol-integrations-are-the-risks-of-external-contracts-pausing-or-executing-an-emergency-withdrawal-acceptable-if-not-watsons-will-submit-issues-related-to-these-situations-that-can-harm-your-protocols-functionality)
> Q: In case of external protocol integrations, are the risks of external contracts pausing or executing an emergency withdrawal acceptable? If not, Watsons will submit issues related to these situations that can harm your protocol's functionality.
> Yes.

**Comment by Oot2k** [comment]
Escalate

I belief this is a valid issue based on comments from Akshay above.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> I belief this is a valid issue based on comments from Akshay above.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Shogoki** [comment]
> * The only possible reason for which an admin would want to replace the existing `token` would be the case when the existing token starts behaving incorrectly (due to bug or hack).

From my PoV this is not the only reason. The Distribution contract could potentially be reused for another distribution if one is finished. (change the token & the Merkle Root)

I think this can be a valid low issue. If the old token is buggy and blocking the contract from being reconfigured, the recovery is relatively easy. 
The admin might just discard this contract and deploy a new one to handle the distribution.

**Comment by akshaysrivastav** [comment]
Hey @Shogoki I will need to respectfully disagree with you on this.
> From my PoV this is not the only reason. The Distribution contract could potentially be reused for another distribution if one is finished. (change the token & the Merkle Root)

This is not true. The [storage state](https://github.com/sherlock-audit/2023-06-tokensoft/blob/main/contracts/contracts/claim/abstract/Distributor.sol#L81-L82) of Distributor contract is never reset and there is no way to reset state via admin as well. So using the same Distributor contract for another token distribution will be a big mistake and can cause incorrect distribution or loss of tokens.

 > If the old token is buggy and blocking the contract from being reconfigured, the recovery is relatively easy.
 
 I would say this recovery mechanism is true for every issue in this contest, the admin can always use `sweepToken` function to recover from [all issues](https://github.com/sherlock-audit/2023-06-tokensoft-judging/issues) that are currently considered as valid.
 
 Again, for the context provided in [this](https://github.com/sherlock-audit/2023-06-tokensoft-judging/issues/146#issuecomment-1671232256) comment, I believe this should be considered as a valid medium.

**Comment by cr-walker** [comment]
I agree with @akshaysrivastav that we shouldn't reuse distributors due to the distributor records (it might be possible but it would definitely be confusing). But since this potential issue depends on a hypothetical problem in a third party contract and there are easy remediations, I don't see it as valid.

Scenario:
* I deploy a distributor to distribute token A
* Oops! This was incorrect configuration, the distributor should be distributing token B
* I call `_setToken()` and it fails due to a bug in token A's implementation
* No problem: just deploy another distributor and sweep the buggy token A out of the old distributor so I can return it

In short, I can't think of a scenario where something catastrophic happens unless the token being distributed is hopelessly broken (i.e. cannot be swept). And if it is, that's not my problem.

**Comment by akshaysrivastav** [comment]
The issue is not about a misconfiguration by admin as described in your example, it is more about the old token preventing the replacement of itself with a new token. These kinds of issues were made `in-scope` by the sponsors ([here](https://github.com/sherlock-audit/2023-06-tokensoft-judging/issues/146#issuecomment-1671232256)).

Again, the sweep token recovery mechanism can be used to recover from any bug in TokenSoft protocol, essentially making all the currently accepted issues as of `low` severity.

**Comment by hrishibhat** [comment]
Result:
Low
Unique
Agree with @Shogoki and Sponsor comments that this  'potential issue depends on a hypothetical problem in a third party contract'
Additionally, 
> Q: In case of external protocol integrations, are the risks of external contracts pausing or executing an emergency withdrawal acceptable? If not, Watsons will submit issues related to these situations that can harm your protocol's functionality.
Yes.

This means that the risks of external contracts are acceptable.

this is a low-impact issue.

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [Oot2k](https://github.com/sherlock-audit/2023-06-tokensoft-judging/issues/146/#issuecomment-1671303699): rejected
